File: app/src/main/java/me/ykrank/s1next/data/api/model/PmGroup.java
package me.ykrank.s1next.data.api.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.Objects;

import me.ykrank.s1next.data.SameItem;

/**
 * Created by ykrank on 2016/11/12 0012.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public final class PmGroup implements Cloneable, SameItem {
    /**
     * who create this pm list
     */
    @JsonProperty("authorid")
    private String authorId;
    /**
     * pm list id
     */
    @JsonProperty("plid")
    private String plId;
    @JsonIgnore
    private boolean isNew;
    @JsonProperty("lastauthorid")
    private String lastAuthorid;
    @JsonProperty("lastauthor")
    private String lastAuthor;
    @JsonProperty("lastsummary")
    private String lastSummary;
    @JsonProperty("lastdateline")
    private long lastDateline;
    @JsonProperty("pmnum")
    private String pmNum;
    @JsonProperty("touid")
    private String toUid;
    @JsonProperty("tousername")
    private String toUsername;

    public PmGroup() {
    }

    @JsonCreator
    public PmGroup(@JsonProperty("isnew") String isNew) {
        this.isNew = "1".equals(isNew);
    }

    public String getAuthorId() {
        return authorId;
    }

    public void setAuthorId(String authorId) {
        this.authorId = authorId;
    }

    public String getPlId() {
        return plId;
    }

    public void setPlId(String plId) {
        this.plId = plId;
    }

    public boolean getIsNew() {
        return isNew;
    }

    public void setIsNew(boolean isNew) {
        this.isNew = isNew;
    }

    public String getLastAuthorid() {
        return lastAuthorid;
    }

    public void setLastAuthorid(String lastAuthorid) {
        this.lastAuthorid = lastAuthorid;
    }

    public String getLastAuthor() {
        return lastAuthor;
    }

    public void setLastAuthor(String lastAuthor) {
        this.lastAuthor = lastAuthor;
    }

    public String getLastSummary() {
        return lastSummary;
    }

    public void setLastSummary(String lastSummary) {
        this.lastSummary = lastSummary;
    }

    public long getLastDateline() {
        return lastDateline;
    }

    public void setLastDateline(long lastDateline) {
        this.lastDateline = lastDateline;
    }

    public String getPmNum() {
        return pmNum;
    }

    public void setPmNum(String pmNum) {
        this.pmNum = pmNum;
    }

    public String getToUid() {
        return toUid;
    }

    public void setToUid(String toUid) {
        this.toUid = toUid;
    }

    public String getToUsername() {
        return toUsername;
    }

    public void setToUsername(String toUsername) {
        this.toUsername = toUsername;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        PmGroup pmGroup = (PmGroup) o;
        return Objects.equal(authorId, pmGroup.authorId) &&
                Objects.equal(plId, pmGroup.plId) &&
                Objects.equal(isNew, pmGroup.isNew) &&
                Objects.equal(lastAuthorid, pmGroup.lastAuthorid) &&
                Objects.equal(lastAuthor, pmGroup.lastAuthor) &&
                Objects.equal(lastSummary, pmGroup.lastSummary) &&
                Objects.equal(lastDateline, pmGroup.lastDateline) &&
                Objects.equal(pmNum, pmGroup.pmNum) &&
                Objects.equal(toUid, pmGroup.toUid) &&
                Objects.equal(toUsername, pmGroup.toUsername);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(authorId, plId, isNew, lastAuthorid, lastAuthor, lastSummary, lastDateline, pmNum, toUid, toUsername);
    }

    @Override
    public boolean isSameItem(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        PmGroup pmGroup = (PmGroup) o;
        return Objects.equal(plId, pmGroup.plId);
    }
}

File: app/src/main/java/me/ykrank/s1next/data/api/model/PmGroup.kt
package me.ykrank.s1next.data.api.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.ykrank.androidtools.ui.adapter.StableIdModel
import com.github.ykrank.androidtools.ui.adapter.model.DiffSameItem

/**
 * Created by ykrank on 2016/11/12 0012.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
class PmGroup : Cloneable, DiffSameItem, StableIdModel {
    /**
     * who create this pm list
     */
    @JsonProperty("authorid")
    var authorId: String? = null
    /**
     * pm list id
     */
    @JsonProperty("plid")
    var plId: String? = null
    @JsonIgnore
    var isNew: Boolean = false
    @JsonProperty("lastauthorid")
    var lastAuthorid: String? = null
    @JsonProperty("lastauthor")
    var lastAuthor: String? = null
    @JsonProperty("lastsummary")
    var lastSummary: String? = null
    @JsonProperty("lastdateline")
    var lastDateline: Long = 0
    @JsonProperty("pmnum")
    var pmNum: String? = null
    @JsonProperty("touid")
    var toUid: String? = null
    @JsonProperty("tousername")
    var toUsername: String? = null

    constructor() {}

    @JsonCreator
    constructor(@JsonProperty("isnew") isNew: String) {
        this.isNew = "1" == isNew
    }

    override val stableId: Long
        get() = plId?.toLongOrNull() ?: 0


    override fun isSameItem(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PmGroup
        if (plId != other.plId) return false

        return true
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PmGroup

        if (authorId != other.authorId) return false
        if (plId != other.plId) return false
        if (isNew != other.isNew) return false
        if (lastAuthorid != other.lastAuthorid) return false
        if (lastAuthor != other.lastAuthor) return false
        if (lastSummary != other.lastSummary) return false
        if (lastDateline != other.lastDateline) return false
        if (pmNum != other.pmNum) return false
        if (toUid != other.toUid) return false
        if (toUsername != other.toUsername) return false

        return true
    }

    override fun hashCode(): Int {
        var result = authorId?.hashCode() ?: 0
        result = 31 * result + (plId?.hashCode() ?: 0)
        result = 31 * result + isNew.hashCode()
        result = 31 * result + (lastAuthorid?.hashCode() ?: 0)
        result = 31 * result + (lastAuthor?.hashCode() ?: 0)
        result = 31 * result + (lastSummary?.hashCode() ?: 0)
        result = 31 * result + lastDateline.hashCode()
        result = 31 * result + (pmNum?.hashCode() ?: 0)
        result = 31 * result + (toUid?.hashCode() ?: 0)
        result = 31 * result + (toUsername?.hashCode() ?: 0)
        return result
    }
}
